### PR TITLE
Prevent use of constant challenge value

### PR DESCRIPTION
**Pull request by selfissued**
... as discussed with @aaronpk and @danielfett  at IETF 118.

### CODE REVIEW

**Comment by panva** [comment]
I am surprised to see yet another attempt to add a measure that requires state synchronization across the deployment.

I wish there was more language with maybe a recommendations about how vigorous this mechanism should be, at which step it should be checking to remain effective, etc.

**Comment by jogu** [comment]
I agree with Filip. I don't think it's necessary to detect every case of re-use. This is really a measure to stop client developers deploying clients into production systems that have poor security, and a 'best efforts' style detection that will eventually stop some requests from clients that are reusing nonces/PKCE challenges is probably okay.

I also think it should not be explicit about it being the PKCE challenge - some authorization servers may prefer to detect reuse of code_verifier instead (they're almost equivalent statements, but referring to PKCE challenges may imply that the detection should be at the authorization endpoint).

**Comment by selfissued** [comment]
What if I changed the language to:
> Authorization servers SHOULD detect and prevent the use of constant PKCE challenge or OpenID Connect `nonce` values.

Then synchronization isn't required.

I believe we need to say something about this possibility, which actually occurs in practice.  If you have different language in mind, please propose text.

**Comment by bc-pi** [comment]
Perhaps something like this might be a more realistic statement and be less objectionable?

> Authorization servers are encouraged to make a reasonable effort at detecting and preventing the use of constant PKCE challenge or OpenID Connect nonce values.

File: recommendations.md [hunk 1]
```diff
@@ -67,6 +67,8 @@ injection attacks (see (#code_injection)) and misuse of authorization codes usin
 In any case, the PKCE challenge or OpenID Connect `nonce` MUST be
 transaction-specific and securely bound to the client and the user agent in
 which the transaction was started.
+Authorization servers MUST detect and prevent the use of constant
+PKCE challenge or OpenID Connect `nonce` values.
```
Comment (selfissued) [inline recommendations.md:71]:
```suggestion
Authorization servers SHOULD detect and prevent the use of constant PKCE challenge or OpenID Connect `nonce` values.
```

**Comment by selfissued** [comment]
I applied the suggested wording to the PR.

**Review by bc-pi — APPROVED** [review]

**Review by panva — APPROVED** [review]

**Review by panva — CHANGES_REQUESTED** [review]
Ah no, it seems brian's suggestion is not the one in the PR at the moment.

File: recommendations.md [hunk 2]
```diff
@@ -67,6 +67,7 @@ injection attacks (see (#code_injection)) and misuse of authorization codes usin
 In any case, the PKCE challenge or OpenID Connect `nonce` MUST be
 transaction-specific and securely bound to the client and the user agent in
 which the transaction was started.
+Authorization servers SHOULD detect and prevent the use of constant PKCE challenge or OpenID Connect `nonce` values.
```
Comment (bc-pi) [inline recommendations.md:70]:
```suggestion
Authorization servers are encouraged to make a reasonable effort at detecting and
preventing the use of constant PKCE challenge or OpenID Connect nonce values.
```

**Review by bc-pi — CHANGES_REQUESTED** [review]

**Comment by bc-pi** [comment]
> Ah no, it seems brian's suggestion is not the one in the PR at the moment. It is in my opinion preferable.

Thanks for catching that Filip. I'd overlooked it in my flu induced and sleep deprived stupor.

File: recommendations.md [hunk 3]
```diff
@@ -67,6 +67,8 @@ injection attacks (see (#code_injection)) and misuse of authorization codes usin
 In any case, the PKCE challenge or OpenID Connect `nonce` MUST be
 transaction-specific and securely bound to the client and the user agent in
 which the transaction was started.
+Authorization servers are encouraged to make a reasonable effort at detecting and
+preventing the use of constant PKCE challenge or OpenID Connect nonce values.
```
Comment (selfissued) [inline recommendations.md:71]:
```suggestion
preventing the use of constant PKCE challenge or OpenID Connect `nonce` values.
```

**Comment by selfissued** [comment]
I've applied Brian's wording suggestion.

**Review by bc-pi — APPROVED** [review]

**Review by panva — APPROVED** [review]

**Review by danielfett — APPROVED** [review]

**Pull request merged**